src/dataflow/solvers/general/DataflowGraphSolvingStrategy.java

```diff
@@ -98,50 +99,65 @@ protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Const
     }
 
     @Override
-    protected InferenceSolution mergeSolution(List<Map<Integer, AnnotationMirror>> inferenceSolutionMaps) {
-        Map<Integer, AnnotationMirror> result = new HashMap<>();
+    protected InferenceResult mergeInferenceResults(List<Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>>> inferenceResults) {
+        Map<Integer, AnnotationMirror> solutions = new HashMap<>();
         Map<Integer, Set<AnnotationMirror>> dataflowResults = new HashMap<>();
-
-        for (Map<Integer, AnnotationMirror> inferenceSolutionMap : inferenceSolutionMaps) {
-            for (Map.Entry<Integer, AnnotationMirror> entry : inferenceSolutionMap.entrySet()) {
-                Integer id = entry.getKey();
-                AnnotationMirror dataflowAnno = entry.getValue();
-                if (AnnotationUtils.areSameByClass(dataflowAnno, DataFlow.class)) {
-                    Set<AnnotationMirror> datas = dataflowResults.get(id);
-                    if (datas == null) {
-                        datas = AnnotationUtils.createAnnotationSet();
-                        dataflowResults.put(id, datas);
+        Set<Constraint> explanations = new HashSet<>();
+
+        for (Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>> inferenceResult : inferenceResults) {
+            Map<Integer, AnnotationMirror> inferenceSolutionMap = inferenceResult.fst;
+            if (inferenceResult.fst != null) {
+                for (Map.Entry<Integer, AnnotationMirror> entry : inferenceSolutionMap.entrySet()) {
+                    Integer id = entry.getKey();
+                    AnnotationMirror dataflowAnno = entry.getValue();
+                    if (AnnotationUtils.areSameByClass(dataflowAnno, DataFlow.class)) {
+                        Set<AnnotationMirror> datas = dataflowResults.get(id);
+                        if (datas == null) {
+                            datas = AnnotationUtils.createAnnotationSet();
+                            dataflowResults.put(id, datas);
+                        }
+                        datas.add(dataflowAnno);
                     }
-                    datas.add(dataflowAnno);
                 }
+            } else {
+                solutions = null;
```

Comment:
Directly return here would make code easier and simpler to read. I.e. you can remove the if condition on line 130, and simplify the return statements on 157-161.